src/App/Fossa/FossaAPIV1.hs

```diff
@@ -580,3 +587,100 @@ resolveProjectDependencies apiOpts locator = fossaReq $ do
   (dependencies :: [ResolvedDependency]) <- responseBody <$> req GET (resolveProjectDependenciesEndpoint baseUrl locator) NoReqBody jsonResponse opts
 
   pure $ map unwrapResolvedDependency dependencies
+
+----------
+
+baseVsiUrl :: Url scheme -> Url scheme
+baseVsiUrl baseurl = baseurl /: "api" /: "proxy" /: "sherlock"
+
+data VSICreateScanRequestBody = VSICreateScanRequestBody
+  { vsiCreateScanRequestBodyOrgID :: Int
+  , vsiCreateScanRequestBodyProjectID :: Text
+  , vsiCreateScanRequestBodyRevisionID :: Text
+  }
+
+instance ToJSON VSICreateScanRequestBody where
+  toJSON VSICreateScanRequestBody{..} =
+    object
+      [ "OrganizationID" .= vsiCreateScanRequestBodyOrgID
+      , "ProjectID" .= vsiCreateScanRequestBodyProjectID
+      , "RevisionID" .= vsiCreateScanRequestBodyRevisionID
+      ]
+
+newtype VSICreateScanResponseBody = VSICreateScanResponseBody {unVSICreateScanResponseBody :: VSI.ScanID}
+
+instance FromJSON VSICreateScanResponseBody where
+  parseJSON = withObject "VSICreateScanResponseBody" $ \obj -> VSICreateScanResponseBody <$> obj .: "ScanID"
+
+vsiCreateScanEndpoint :: Url scheme -> Url scheme
+vsiCreateScanEndpoint baseurl = baseVsiUrl baseurl /: "scans"
+
+vsiCreateScan :: (Has (Lift IO) sig m, Has Diagnostics sig m) => ApiOpts -> ProjectRevision -> m VSI.ScanID
+vsiCreateScan apiOpts ProjectRevision{..} = fossaReq $ do
+  (baseUrl, baseOpts) <- useApiOpts apiOpts
+
+  Organization orgId _ <- getOrganization apiOpts
+  let projectID = renderLocator $ Locator "custom" projectName Nothing
+  let reqBody = VSICreateScanRequestBody orgId projectRevision projectID
+
+  (resBody :: VSICreateScanResponseBody) <- responseBody <$> req POST (vsiCreateScanEndpoint baseUrl) (ReqBodyJson reqBody) jsonResponse baseOpts
+  pure $ unVSICreateScanResponseBody resBody
+
+newtype VSIAddFilesToScanRequestBody = VSIAddFilesToScanRequestBody {vsiAddFilesToScanRequestBodyFiles :: Map (Path Rel File) Fingerprint.Combined}
+
+instance ToJSON VSIAddFilesToScanRequestBody where
+  toJSON VSIAddFilesToScanRequestBody{..} = object ["ScanData" .= vsiAddFilesToScanRequestBodyFiles]
+
+vsiAddFilesToScanEndpoint :: Url scheme -> VSI.ScanID -> Url scheme
+vsiAddFilesToScanEndpoint baseurl scanID = baseVsiUrl baseurl /: "scans" /: toText scanID /: "files"
```

Comment:
Using a `toText` here kind of breaks the type safety that we gain with the `ScanID` newtype.
I would usually prefer to see something like this:
```suggestion
vsiAddFilesToScanEndpoint baseurl (VSI.ScanID scanID) = baseVsiUrl baseurl /: "scans" /: scanID /: "files"
```

Reply:
Fair enough, I buy that argument. Thanks!